src/Fixer/ClassNotation/FinalClassFixer.php

```diff
@@ -0,0 +1,113 @@
+<?php
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Fixer\ClassNotation;
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\DocBlock\DocBlock;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+/**
+ * @author Filippo Tessarotto <zoeslam@gmail.com>
+ */
+final class FinalClassFixer extends AbstractFixer
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getDefinition()
+    {
+        return new FixerDefinition(
+            'All classes must be final, except abstract ones and Doctrine entities.',
```

Comment:
one more case before we will be able to merge it
classes marked with `@final` annotation shall not be marked with `final` keyword.

eg, if PHP CS Fixer v2 has some class that is not final, fixer would make it final, but that's BC breaker, so we would move `final` keyword into `@final` annotation and plan to make it real final in v3.

Reply:
This would be against the purpose of the fixer.

`@final` is not a standard annotation (AFAIK), and if you have a non trivial DocBlock for the class (more than 10 lines) it would be very easy to miss out the tag (in a contributor PR f.e.) and therefore the class could end up without the final php keyword.

Being a BC breaker is intentional. I want maintainers to expect that adopting this fixer automatically prevent each and every inheritance (future) break.

As a maintainer myself, I rely on (risky) fixers also to spot such dangerous cases.

Reply:
there is no docblock PSR approved you know? :)
yet, this annotation is highly used.

I totally agree of relying on rule as a watcher to not introduce new classes without `final`!
Yet, like at PHP CS Fixer, we already have the situation when we have non-final classes, and we can not mark them as final. So, we can either do nothing, which is bad, or mark them with `@final` and mark as real final in next MAJOR release. But doing the later will prevent us from using the rule.

Reply:
On the other side, without it the rule will be used either in green field project or project which already follow the rule.
The rule won't help community to follow the rule if they are not following it yet; or they will start it and do the BC breaks

Reply:
The only two cases that come into my mind are:

1. A project didn't add any DocBlock to state (in the current version) that a class shouldn't be used in the userland, so adding `final` would be a BC break anyway
1. A project, like this, already had a dedicated DocBlock, so `FinalInternalClassFixer` would fit much better for the transition period

With the configuration you ask, it would be much easier to mark `annotation-white-list` option as optional in `FinalInternalClassFixer`, and drop this PR.

Reply:
do you mean that this PR can be basically a proxy fixer to
`final_internal_class => ['annotation-black-list'=> ['@Entity', '@ORM'], 'annotation-white-list ' => []]`
?

Reply:
Yes

Reply:
maybe I'm slow on this one, but if we already have a fixer to this what is the added value of another?

Reply:
That there is no way to escape this fixer. Once you enable it, your library will never have inheritance BC breaks anymore, since subclassing would be totally forbidden for non abstract classes.

On the other hand, `final_internal_class` provide much weaker guarantees, because it relies on DocBlock to take decisions, and DocBlocks are much harder to track/review.

What we could do is proxy this fixer to `final_internal_class => ['annotation-black-list'=> ['@Entity', '@ORM'], 'annotation-white-list ' => []]` to reduce code duplication.

Reply:
that can make sense.
ultimately, new code from this PR is nice and test cases brings a lot of value.

i would suggest to merge as much logic & tests as possible into `final_internal_class`, and then to the proxy-ing

then, I would actually rename `final_internal_class` into `final_class` and change default config on it on 3.0, as final by default is better, and exception of `@internal`/`@final` are ways of dealing with legacy code and looks more like a configuration per project from perspective.

cc @Slamdunk @SpacePossum

Reply:
I really like the suggestion of @keradus here, TBH I think it is a requirement for this PR. This is because without it this PR would bring a mixture of a proxy fixer with new logic which might cause confusion.
So lets aim for `final` as much as possible with the new fixer proposed here and allow for the detailed exceptions of the current fixer, win-win :)

Reply:
Sorry for the late edit. I've proxified the code to `FinalInternalClassFixer` as agreed.